File: front/src/game/stages.ts
import type { Block, BlockType, Cell, Direction, IoProblem, Puzzle, Stage } from './types';
import bundled from './stages.data.json';

/** データファイル上の模範解答ブロック(id なし。ネストは body で表現) */
export interface RawSolutionBlock {
  kind: BlockType;
  times?: number;
  body?: RawSolutionBlock[];
}

/** バックエンド/生成データの転送フォーマット(rows は '#'=壁 '.'=床 'G'=ゴール) */
export interface StageJson {
  id: string;
  name: string;
  mode: 'block' | 'code';
  rows: string[];
  start: { x: number; y: number; dir: Direction };
  allowedBlocks: BlockType[];
  starThresholds: [number, number];
  maxSteps: number;
  hint?: string;
  puzzle?: Puzzle;
  template?: string;
  statement?: string;
  io?: IoProblem;
  solution?: string;
  solutionBlocks?: RawSolutionBlock[];
}

let solutionBlockSeq = 0;

/** id なしの模範解答ブロック列に、再帰的に一意な id を振る */
function assignSolutionBlockIds(raw: RawSolutionBlock[]): Block[] {
  return raw.map((b) => {
    solutionBlockSeq++;
    const id = `sol${solutionBlockSeq}`;
    return b.body
      ? { id, kind: b.kind, times: b.times, body: assignSolutionBlockIds(b.body) }
      : { id, kind: b.kind, times: b.times };
  });
}

export function toStage(j: StageJson): Stage {
  return {
    id: j.id,
    name: j.name,
    mode: j.mode,
    grid: j.rows.map((row) =>
      [...row].map((ch): Cell => (ch === '#' ? 'wall' : ch === 'G' ? 'goal' : 'floor')),
    ),
    start: j.start,
    allowedBlocks: j.allowedBlocks,
    starThresholds: j.starThresholds,
    maxSteps: j.maxSteps,
    hint: j.hint,
    puzzle: j.puzzle,
    template: j.template,
    statement: j.statement,
    io: j.io,
    solution: j.solution,
    solutionBlocks: j.solutionBlocks ? assignSolutionBlockIds(j.solutionBlocks) : undefined,
  };
}

/**
 * フロント同梱のステージデータ(scripts/gen-stages.ts が生成)。
 * 通常はバックエンド API から取得し、API が使えないときのフォールバックとして使う。
 */
export const BUNDLED_STAGES: Stage[] = (bundled as StageJson[]).map(toStage);

export const CODE_TEMPLATE = `#include "game.h"

// つかえる関数:
//   move_forward();    // 1マスすすむ
//   turn_left();       // ひだりをむく
//   turn_right();      // みぎをむく
//   is_wall_ahead();   // まえがかべなら1
//   is_goal();         // ゴールなら1

int main(void) {
    // ここに プログラムを かこう

    return 0;
}
`;

File: back/src/stages.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// ステージデータ。data/stages.json(front/scripts/gen-stages.ts が生成)と同スキーマ。
#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Stage {
    pub id: String,
    pub name: String,
    pub mode: String,
    /// '#'=壁 '.'=床 'G'=ゴール
    pub rows: Vec<String>,
    pub start: Start,
    pub allowed_blocks: Vec<String>,
    pub star_thresholds: [u32; 2],
    pub max_steps: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hint: Option<String>,
    /// 配列パズル(ソート等)のみ
    #[serde(skip_serializing_if = "Option::is_none")]
    pub puzzle: Option<Puzzle>,
    /// ステージ固有のコードテンプレート
    #[serde(skip_serializing_if = "Option::is_none")]
    pub template: Option<String>,
    /// 問題文本文(上級のみ。IO問題では必須)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub statement: Option<String>,
    /// 標準入出力問題(paiza/AtCoder 風)。ある場合、盤面・トレース再生は使わない
    #[serde(skip_serializing_if = "Option::is_none")]
    pub io: Option<IoProblem>,
    /// 模範解答の C コード(上級のみ)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub solution: Option<String>,
    /// 模範解答のブロック列(初級のみ。パススルー配信、実行はフロントのインタプリタ)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub solution_blocks: Option<Vec<SolutionBlock>>,
}

/// 初級の模範解答ブロック(id はフロントのロード時に付与される)
#[derive(Clone, Serialize, Deserialize)]
pub struct SolutionBlock {
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub times: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<Vec<SolutionBlock>>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct Puzzle {
    pub kind: String, // sort | sortDesc | reverse | maxLast | minFirst
    pub values: Vec<i32>,
}

/// 標準入出力問題(paiza/AtCoder 風)。front/src/game/types.ts の IoProblem と同スキーマ。
#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IoProblem {
    pub input_format: String,
    pub output_format: String,
    pub constraints: String,
    pub samples: Vec<IoSample>,
    pub hidden_tests: Vec<IoCaseSpec>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct IoSample {
    pub input: String,
    pub output: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub note: Option<String>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct IoCaseSpec {
    pub input: String,
    pub output: String,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct Start {
    pub x: i32,
    pub y: i32,
    pub dir: String,
}

/// ビルド時に data/stages.json を同梱して読み込む
pub fn load() -> Vec<Stage> {
    let json = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/../data/stages.json"));
    serde_json::from_str(json).expect("data/stages.json is invalid")
}

pub fn index_by_id(stages: &[Stage]) -> HashMap<String, Stage> {
    stages.iter().map(|s| (s.id.clone(), s.clone())).collect()
}
